docs/secure-boot.md

```diff
@@ -1,39 +1,172 @@
+* The `balena-boot` partition is encrypted and contains everything else, as these files may contain secrets such as passwords or API keys which the encryption should protect.
+
+The partitions are mounted under `/mnt/boot` and `/mnt/efi` respectively and `/mnt/boot/EFI` is symlinked to `../efi/EFI` to make sure all the paths are still valid.
+
+## Using linux kernel as a 2nd stage bootloader
+
+In order to avoid the need to make GRUB use the TPM to unlock the partitions when looking for the kernel, we keep an unencrypted copy of the linux kernel in the EFI partition whose sole purpose is to unlock the root partition using the TPM, find the actual kernel and kexec into it. This is a simple version of a completely different feature but it roughly works as follows:
+1. During provisioning the installer copies an instance of its own/currently running kernel into the EFI partition.
+2. When the newly provisioned OS boots, UEFI loads GRUB.
+3. GRUB finds the unencrypted 2nd stage kernel in the EFI partition and executes it adding `balena_stage2` parameter to the kernel command line.
+4. The attached initramfs uses the TPM to unlock the root partition.
+5. Next the initramfs checks whetner the `balena_stage2` parameter is present and if it is, loads the actual kernel from the root partition and kexec's into it, removing the `balena_stage2` parameter from the kernel command line and replacing the `root` parameter by the actual UUID (which is only available after unlocking LUKS).
+6. The actual kernel from the root partition is executed, boot continues as usual.
+
+Though not currently implemented, the 2nd stage kernel in the EFI partition is intended to be replaced by currently running kernel on successful HUP.
+
+## Updates
+
+### OS updates (HUP)
 
-### Setup mode
+Each OS build ships update files for `db` and `dbx` in the `/resin-boot` directory. These contain the hashes of GRUB to make them bootable. The files are in `.auth` format - they are signed by KEK and applicable from runtime. The OS update process is:
+1. New hostapp is downloaded and unpacked to the inactive partition
+2. Hashes from the new `db.auth` file are appended to the `db` variable. This makes both the current and new OS's bootable.
+3. Device reboots, new OS boots up.
+4. `rollback-health` checks whether the new OS is healthy.
+5. If the update failed, rollback triggers and the old OS is booted up. Both OS's have valid signatures in `db` and both are still bootable so that the update can be retried.
+6. If the update went through fine, `rollback-health` will look for `dbx.auth` in the inactive partition and append it to the `dbx` variable. Theoretically the old OS is no longer bootable, however this is only the case if the GRUB binary has actually changed, otherwise its hash in `db` will stay the same and both the old and the new OS's stay bootable.
```

Comment:
> this is only the case if the GRUB binary has actually changed

Wondering whether we should forcefully change Grub to guarantee that older versions are not bootable.

Reply:
I was thinking about this and I think we should consider generating a new GPG key for each GRUB build. We do not really need the key to be persistent, we just need the combination of GRUB+kernel in each hostapp to be bootable. I don't think it would be a problem to change the GRUB GPG key every time so that the GRUB+kernel combo is unique for each build. HUP and rollback should always update both or neither anyway.

Open to wider discussion, we can either do this using the balena-sign API or completely locally (similar to the kernel's built-in modules signing).

Reply:
agreed @mtoman do you want to add a task to the improvement and explore this path?

Reply:
https://balena.fibery.io/search/uVuoN#Tasks/Task/Explore-the-idea-of-generating-a-new-GRUB-GPG-key-for-each-build-575